File: c/p387.c
#include <stdio.h>

#include "util.h"

const long LIMIT = 100000000000000;

long right_strong_prime_sum(long n, long sum);

int main(int argc, char** argv) {
    long sum = 0;
    for (int d = 1; d < 10; ++d) {
        sum += right_strong_prime_sum(d, d);
    }
    printf("%ld\n", sum);
}

long right_strong_prime_sum(const long n, const long digit_sum) {
    long total = 0;
    for (int d = 0; d < 10; ++d) {
        const long next = n * 10 + d;
        const long next_digit_sum = digit_sum + d;

        if (next > LIMIT) {
            return total;
        }

        if (next % next_digit_sum == 0) { // right
            if (is_prime(next / next_digit_sum)) { // strong
                for (int dd = 1; dd < 10; dd += 2) {
                    long nn = next * 10 + dd;
                    if (is_prime(nn) && nn < LIMIT) {
                        total += nn;
                    }
                }
            }
            total += right_strong_prime_sum(next, next_digit_sum);
        }
    }
    return total;
}

File: erlang/src/p387.erl
-module(p387).

-export([start/0]).

-define(LIMIT, 100000000000000).

f() ->
  lists:sum([g(D, D) || D <- lists:seq(1, 9)]).

g(N, _DigitSum) when N > ?LIMIT -> 0;
g(N, DigitSum) -> g(N, 0, DigitSum).

g(_N, 10, _DigitSum) -> 0;
g(N, I, DigitSum) ->
  Next = N * 10 + I,
  NextDigitSum = DigitSum + I,
  case right(Next, NextDigitSum) of
    true -> strong_sum(Next, NextDigitSum) + g(Next, NextDigitSum);
    false -> 0
  end + g(N, I + 1, DigitSum).

strong_sum(N, DigitSum) ->
  case utils:is_prime(N div DigitSum) of
    true ->
      lists:sum([N * 10 + D || D <- lists:seq(1, 9, 2), N * 10 + D < ?LIMIT, utils:is_prime(N * 10 + D)]);
    false -> 0
  end.

right(N, DigitSum) ->
  N rem DigitSum == 0.

start() ->
  io:format("~p ~n", [f()]).
